File: src/lib/garage/bikeAssist.ts
import type { BikeCategory, WheelSize } from "@/types";

/** Muskel vs. E-Bike — Parität zu Mobile `bike_assist.dart`. */
export type BikeAssistMode = "muscle" | "ebike";

export const MUSCLE_CATEGORIES: BikeCategory[] = [
  "urban",
  "cargo",
  "folding",
  "kids",
  "gravel",
  "road",
  "mtb_trail",
  "mtb_am",
  "mtb_enduro",
  "dh",
  "hiking",
];

/** UI-Untertypen unter E-Bike (Web persistiert category + isEbike). */
export const EBIKE_CATEGORIES: BikeCategory[] = [
  "etrekking",
  "urban",
  "cargo",
  "folding",
  "kids",
  "gravel",
  "road",
  "emtb",
];

export type CategoryPickGroupId = "everyday" | "tour" | "trail";

export type CategoryPickGroup = {
  id: CategoryPickGroupId;
  label: string;
  categories: BikeCategory[];
};

const EVERYDAY_MUSCLE: BikeCategory[] = ["urban", "cargo", "folding", "kids"];
const EVERYDAY_EBIKE: BikeCategory[] = [
  "etrekking",
  "urban",
  "cargo",
  "folding",
  "kids",
];
const TOUR: BikeCategory[] = ["gravel", "road"];
const TRAIL_MUSCLE: BikeCategory[] = [
  "mtb_trail",
  "mtb_am",
  "mtb_enduro",
  "dh",
  "hiking",
];
const TRAIL_EBIKE: BikeCategory[] = ["emtb"];

/** Kurze Typ-Liste beim Anlegen — Feinschnitt später in der Identität. */
export const ADD_MUSCLE: BikeCategory[] = [
  "urban",
  "gravel",
  "road",
  "mtb_am",
  "cargo",
  "folding",
];

export const ADD_EBIKE: BikeCategory[] = [
  "urban",
  "etrekking",
  "gravel",
  "road",
  "emtb",
  "cargo",
];

export function addCategories(mode: BikeAssistMode): BikeCategory[] {
  return mode === "ebike" ? ADD_EBIKE : ADD_MUSCLE;
}

/** Enduro/Trail/DH aus Onboarding markieren die MTB-Kachel. */
export function addTileSelected(
  tile: BikeCategory,
  current: BikeCategory,
  mode: BikeAssistMode
): boolean {
  if (tile === current) return true;
  if (mode === "muscle" && tile === "mtb_am") {
    return current === "mtb_trail" || current === "mtb_enduro" || current === "dh";
  }
  return false;
}

export function defaultWheelFor(category: BikeCategory): WheelSize {
  if (category === "gravel") return "650b";
  if (
    category === "urban" ||
    category === "road" ||
    category === "etrekking" ||
    category === "cargo" ||
    category === "folding" ||
    category === "kids"
  ) {
    return "700c";
  }
  return "29";
}

export function persistCategory(
  uiCategory: BikeCategory,
  mode: BikeAssistMode
): BikeCategory {
  if (mode === "muscle") {
    if (uiCategory === "emtb") return "mtb_am";
    if (uiCategory === "etrekking") return "urban";
    return uiCategory;
  }
  if (
    uiCategory === "mtb_trail" ||
    uiCategory === "mtb_am" ||
    uiCategory === "mtb_enduro" ||
    uiCategory === "dh"
  ) {
    return "emtb";
  }
  if (uiCategory === "hiking") return "etrekking";
  return uiCategory;
}

/** Anlegen: Alltag zuerst, Trail nicht als Default-Welt. */
export function categoryPickGroups(mode: BikeAssistMode): CategoryPickGroup[] {
  const allowed = new Set(
    mode === "ebike" ? EBIKE_CATEGORIES : MUSCLE_CATEGORIES
  );
  const take = (ids: BikeCategory[]) => ids.filter((c) => allowed.has(c));
  const groups: CategoryPickGroup[] = [
    { id: "everyday", label: "Alltag", categories: take(mode === "ebike" ? EVERYDAY_EBIKE : EVERYDAY_MUSCLE) },
    { id: "tour", label: "Tour", categories: take(TOUR) },
    { id: "trail", label: "Trail", categories: take(mode === "ebike" ? TRAIL_EBIKE : TRAIL_MUSCLE) },
  ];
  return groups.filter((g) => g.categories.length > 0);
}

export function assistModeFor(
  category: BikeCategory,
  isEbike = false
): BikeAssistMode {
  if (isEbike || category === "emtb" || category === "etrekking") {
    return "ebike";
  }
  return "muscle";
}

export function subtypeLabel(
  category: BikeCategory,
  mode: BikeAssistMode
): string {
  if (mode === "muscle") {
    const map: Record<BikeCategory, string> = {
      mtb_trail: "MTB Trail",
      mtb_am: "All-Mountain",
      mtb_enduro: "Enduro",
      dh: "Downhill",
      gravel: "Gravel",
      road: "Rennrad",
      urban: "City",
      cargo: "Lastenrad",
      folding: "Faltrad",
      kids: "Kinderrad",
      emtb: "E-MTB",
      etrekking: "E-Trekking",
      hiking: "Wandern",
    };
    return map[category];
  }
  switch (category) {
    case "emtb":
      return "E-MTB";
    case "etrekking":
      return "E-Trekking";
    case "gravel":
      return "E-Gravel";
    case "urban":
      return "E-City";
    case "cargo":
      return "E-Lastenrad";
    case "folding":
      return "E-Faltrad";
    case "kids":
      return "E-Kinderrad";
    case "road":
      return "E-Road";
    default:
      return category;
  }
}

export function coerceCategory(
  current: BikeCategory,
  mode: BikeAssistMode
): BikeCategory {
  if (mode === "muscle") {
    if (current === "emtb") return "mtb_am";
    if (current === "etrekking") return "urban";
    return MUSCLE_CATEGORIES.includes(current) ? current : "urban";
  }
  if (
    current === "mtb_trail" ||
    current === "mtb_am" ||
    current === "mtb_enduro" ||
    current === "dh"
  ) {
    return "emtb";
  }
  if (current === "hiking") return "etrekking";
  if (EBIKE_CATEGORIES.includes(current)) return current;
  return "urban";
}

export function persistIsEbike(
  category: BikeCategory,
  mode: BikeAssistMode
): boolean {
  if (mode === "ebike") return true;
  return category === "emtb" || category === "etrekking";
}

File: mobile/lib/domain/bike_assist.dart
import 'bike.dart';
import 'catalog_bike.dart';

/// Muskel vs. E-Bike — UI-Steuerung für Garage (Mobile/Web-Parität).
enum BikeAssistMode { muscle, ebike }

/// Disziplin-Untertypen je Assist-Modus.
///
/// Mobile persistiert wie Web: [Bike.category] + [Bike.isEbike]
/// (Drift-Spalte `is_ebike`). E-City/E-Gravel/E-Road bleiben `urban`/`gravel`/
/// `road` mit `isEbike=true` — kein Kollaps auf `etrekking`.
class BikeAssistUx {
  BikeAssistUx._();

  static const muscleCategories = <BikeCategory>[
    BikeCategory.urban,
    BikeCategory.cargo,
    BikeCategory.folding,
    BikeCategory.kids,
    BikeCategory.gravel,
    BikeCategory.road,
    BikeCategory.mtbTrail,
    BikeCategory.mtbAm,
    BikeCategory.mtbEnduro,
    BikeCategory.dh,
    BikeCategory.hiking,
  ];

  /// UI-Untertypen unter E-Bike (1:1 mit Persistenz + isEbike).
  static const ebikeCategories = <BikeCategory>[
    BikeCategory.etrekking,
    BikeCategory.urban,
    BikeCategory.cargo,
    BikeCategory.folding,
    BikeCategory.kids,
    BikeCategory.gravel,
    BikeCategory.road,
    BikeCategory.emtb,
  ];

  static const everydayMuscle = <BikeCategory>[
    BikeCategory.urban,
    BikeCategory.cargo,
    BikeCategory.folding,
    BikeCategory.kids,
  ];

  static const everydayEbike = <BikeCategory>[
    BikeCategory.etrekking,
    BikeCategory.urban,
    BikeCategory.cargo,
    BikeCategory.folding,
    BikeCategory.kids,
  ];

  static const tourCategories = <BikeCategory>[
    BikeCategory.gravel,
    BikeCategory.road,
  ];

  static const trailMuscle = <BikeCategory>[
    BikeCategory.mtbTrail,
    BikeCategory.mtbAm,
    BikeCategory.mtbEnduro,
    BikeCategory.dh,
    BikeCategory.hiking,
  ];

  static const trailEbike = <BikeCategory>[BikeCategory.emtb];

  /// Kurze Typ-Liste beim Anlegen — Feinschnitt später in der Identität.
  static const addMuscle = <BikeCategory>[
    BikeCategory.urban,
    BikeCategory.gravel,
    BikeCategory.road,
    BikeCategory.mtbAm,
    BikeCategory.cargo,
    BikeCategory.folding,
  ];

  static const addEbike = <BikeCategory>[
    BikeCategory.urban,
    BikeCategory.etrekking,
    BikeCategory.gravel,
    BikeCategory.road,
    BikeCategory.emtb,
    BikeCategory.cargo,
  ];

  static List<BikeCategory> addCategories(BikeAssistMode mode) =>
      mode == BikeAssistMode.ebike ? addEbike : addMuscle;

  /// Enduro/Trail/DH aus Onboarding markieren die MTB-Kachel.
  static bool addTileSelected(
    BikeCategory tile,
    BikeCategory current,
    BikeAssistMode mode,
  ) {
    if (tile == current) return true;
    if (mode == BikeAssistMode.muscle && tile == BikeCategory.mtbAm) {
      return current == BikeCategory.mtbTrail ||
          current == BikeCategory.mtbEnduro ||
          current == BikeCategory.dh;
    }
    return false;
  }

  static WheelSize defaultWheelFor(BikeCategory c) => switch (c) {
        BikeCategory.urban ||
        BikeCategory.road ||
        BikeCategory.etrekking ||
        BikeCategory.cargo ||
        BikeCategory.folding ||
        BikeCategory.kids =>
          WheelSize.c700,
        BikeCategory.gravel => WheelSize.b650,
        _ => WheelSize.w29,
      };

  /// Anlegen: Trail zuerst (Katalog-Schwerpunkt). Zu Fuß ist kein Rad.
  static List<({String id, String label, List<BikeCategory> categories})>
      pickGroups(
    BikeAssistMode mode, {
    bool includeHiking = false,
  }) {
    final allowed = mode == BikeAssistMode.ebike
        ? ebikeCategories
        : muscleCategories;
    List<BikeCategory> take(List<BikeCategory> raw) => [
          for (final c in raw)
            if (allowed.contains(c) &&
                (includeHiking || c != BikeCategory.hiking))
              c,
        ];
    return [
      (
        id: 'trail',
        label: 'Trail',
        categories: take(
          mode == BikeAssistMode.ebike ? trailEbike : trailMuscle,
        ),
      ),
      (id: 'tour', label: 'Tour', categories: take(tourCategories)),
      (
        id: 'everyday',
        label: 'Alltag',
        categories: take(
          mode == BikeAssistMode.ebike ? everydayEbike : everydayMuscle,
        ),
      ),
    ].where((g) => g.categories.isNotEmpty).toList();
  }

  static BikeAssistMode modeFor({
    required BikeCategory category,
    bool isEbike = false,
  }) {
    if (isEbike ||
        category == BikeCategory.emtb ||
        category == BikeCategory.etrekking) {
      return BikeAssistMode.ebike;
    }
    return BikeAssistMode.muscle;
  }

  /// Label für die Typ-Kachel (unterscheidet E-Gravel vs. Gravel etc.).
  static String subtypeLabel(BikeCategory category, BikeAssistMode mode) {
    if (mode == BikeAssistMode.muscle) {
      return Bike(id: '', name: '', category: category).categoryLabel;
    }
    return switch (category) {
      BikeCategory.emtb => 'E-MTB',
      BikeCategory.etrekking => 'E-Trekking',
      BikeCategory.gravel => 'E-Gravel',
      BikeCategory.urban => 'E-City',
      BikeCategory.cargo => 'E-Lastenrad',
      BikeCategory.folding => 'E-Faltrad',
      BikeCategory.kids => 'E-Kinderrad',
      BikeCategory.road => 'E-Road',
      BikeCategory.mtbTrail => 'E-MTB Trail',
      BikeCategory.mtbAm => 'E-MTB',
      BikeCategory.mtbEnduro => 'E-Enduro',
      BikeCategory.dh => 'E-DH',
      BikeCategory.hiking => 'Zu Fuß',
    };
  }

  /// Anzeige-Label aus Persistenz (category + isEbike).
  static String displayLabel({
    required BikeCategory category,
    bool isEbike = false,
  }) {
    return subtypeLabel(
      category,
      modeFor(category: category, isEbike: isEbike),
    );
  }

  /// Beim Wechsel Muskel ↔ E-Bike sinnvollen Subtyp wählen.
  static BikeCategory coerceCategory(
    BikeCategory current,
    BikeAssistMode mode,
  ) {
    if (mode == BikeAssistMode.muscle) {
      return switch (current) {
        BikeCategory.emtb => BikeCategory.mtbAm,
        BikeCategory.etrekking => BikeCategory.urban,
        _ => muscleCategories.contains(current)
            ? current
            : BikeCategory.urban,
      };
    }
    return switch (current) {
      BikeCategory.mtbTrail ||
      BikeCategory.mtbAm ||
      BikeCategory.mtbEnduro ||
      BikeCategory.dh =>
        BikeCategory.emtb,
      BikeCategory.hiking => BikeCategory.etrekking,
      _ => current,
    };
  }

  /// Persistenz-Kategorie (Web-Parität): UI-Untertyp behalten.
  ///
  /// E-MTB → `emtb`; E-Trekking → `etrekking`; E-City/E-Gravel/E-Road →
  /// `urban`/`gravel`/`road` (Assist über [persistIsEbike]).
  static BikeCategory persistCategory(
    BikeCategory uiCategory,
    BikeAssistMode mode,
  ) {
    if (mode == BikeAssistMode.muscle) {
      if (uiCategory == BikeCategory.emtb) return BikeCategory.mtbAm;
      if (uiCategory == BikeCategory.etrekking) return BikeCategory.urban;
      return uiCategory;
    }
    return switch (uiCategory) {
      BikeCategory.mtbTrail ||
      BikeCategory.mtbAm ||
      BikeCategory.mtbEnduro ||
      BikeCategory.dh =>
        BikeCategory.emtb,
      BikeCategory.hiking => BikeCategory.etrekking,
      _ => uiCategory,
    };
  }

  /// Explizites Assist-Flag für Persistenz (Web `persistIsEbike`).
  static bool persistIsEbike(
    BikeCategory category,
    BikeAssistMode mode,
  ) {
    if (mode == BikeAssistMode.ebike) return true;
    return category == BikeCategory.emtb || category == BikeCategory.etrekking;
  }

  /// Katalog → Persistenz: Disziplin behalten wenn schon E-encoded;
  /// niemals alles auf `emtb` bzw. `etrekking` zwingen.
  static ({BikeCategory category, bool isEbike}) resolveCatalogPersist(
    CatalogBikeVariant cat,
  ) {
    if (!cat.isEbike) {
      return (category: cat.category, isEbike: false);
    }
    final persisted = persistCategory(cat.category, BikeAssistMode.ebike);
    return (category: persisted, isEbike: true);
  }
}
